### DISCUSSION TITLE
Catch all route takes precedence on reload while dynamic route takes precedence on routing

### DISCUSSION THREAD

**Discussion by wr9** [vercel/next.js#37020] [Help] [answered]
So I have a catch all route and a dynamic route, let's say

pages
--category
----[name].js
--[...parameters].js

The dynamic route, `[name]` in the `category` folder has `getStaticPaths` based on which it generates `/category/A`, `/category/B`, `/category/C` etc. `getStaticProps` returns some data, and if `context.params.name === 'blabla'` returns also `notFound: true`

The catch all route, `[...parameters]` is able to handle `/name/blabla` path and also its `getStaticPaths` returns it.

When I build the app, the build goes fine, `/category/A`, `/category/B` and `/category/blabla` are all generated.

**The issue:**

If I'm loading/reloading the app with the `/category/blabla` route -> all is fine, and I can see that pages/[...parameters], the catch all route is handling this.

If I'm loading/reloading the app with the `/category/A` route -> all is fine, however if I click on a link to `/category/blabla` and am routed to `/category/blabla`, I can see that now `category/[name]` is handling this. In my case my app breaks because `category/[name]` does not know how to handle  `/category/blabla` route. If I reload while on `/category/blabla`, all is fine because it's once again handled by `pages/[...parameters]`, the catch all route.

**So to make it short**
`/category/blabla` => catch all route is handling this
`/category/A` -> `/category/blabla` => dynamic route is handling this

So does anyone know if there is anything that I can do about this? I want `/category/blabla` to always be handled by the catch all route.

Also yes I am aware of the precedence, however the behaviour related to this is obviously inconsistent.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

When you link to `/category/blabla` can you tell somehow that it is invalid, from the client? If that's the case, I think you could use `<Link href="/[...parameters]" as="/category/blabla"/>`, and that should help, but it doesn't feel optimal.  **This worked out as expected**.

~In `/category/[name]`, do you have `GetStaticPaths` with fallback set `true` or `"blocking"`? Trying to build a reproduction demo, or if you have one that'd be great, I am curious about what's going on here.~

```tsx
// category/[name]
import type { GetStaticPaths, GetStaticProps, NextPage } from "next";

const Category: NextPage<{ category: string }> = ({ category }) => {
  return <div>Category: {category}</div>;
};

export const getStaticPaths: GetStaticPaths<{ name: string }> = () => {
  return {
    paths: [{ params: { name: "A" } }, { params: { name: "B" } }],
    fallback: false,
  };
};

export const getStaticProps: GetStaticProps = (ctx) => {
  const name = ctx.params?.name;

  if (!name) return { notFound: true };

  return { props: { category: name } };
};

export default Category;
```

```tsx
// [...parameters]
import type { NextPage } from "next";
import { useRouter } from "next/router";

const Fallback: NextPage = () => {
  const { parameters } = useRouter().query;

  if (!parameters) return <div>No params</div>;

  return <div>Fallback: {[parameters].join(", ")}</div>;
};

export default Fallback;
```

```tsx
// index
import type { NextPage } from "next";
import Link from "next/link";

const Home: NextPage = () => {
  return (
    <section>
      <h1>Home</h1>

      <ul>
        <li>
          <Link href="/category/A">A</Link>
        </li>
        <li>
          <Link href="/category/B">B</Link>
        </li>
        <li>
          <Link href="/[...parameters]" as="/category/C">
            C
          </Link>
        </li>
      </ul>
    </section>
  );
};

export default Home;
```

The above combination works, but yeah, it relies on the `href/as` combo, and it requires you to know upfront that `C` is not an allowed route.

**Reply by wr9** [reply]
Works for my case - this is excellent, thank you. So simple yet I've missed it.

In my case, I have `fallback: 'blocking` in `getStaticPaths` for `category/[name]` but I do know in it that `/category/blabla` is not an allowed route so I return `notFound: true` for that route, in `getStaticProps`.
